README.md

```diff
@@ -189,6 +189,13 @@ jobs:
 
 **Note**: the `otp-version: false` input is only applicable when installing Gleam.
 
+## Hex.pm mirrors and caching
+
+It is possible to define the hex.pm mirrors to use including their order with
+option `hexpm-mirrors`. By default, the action will fetch from `repo.hex.pm`
+first and then from `cdn.jsdelivr.net/hex`. Permanent files from these mirrors,
```

Comment:
This should not be a default, cdn.jsdelivr.net/hex is not an official mirror and we cannot give an any guarantees for it. Unlike when download Hex packages there is no public key verification when downloading builds. It is also known to sometimes lag behind by a few days on certain paths.

I think retries are a great idea but should be done against the official repo and if users want to use a different repo or multiple with round robin then that should be opt in.

Reply:
Thank you for your review.

I am confused about these two points:

> Unlike when download Hex packages there is no public key verification when downloading builds.

Are you referring to the mechanism in [Mix.Local](https://github.com/elixir-lang/elixir/blob/d9c7e80302c220a2b603ac16931dee6056a6c863/lib/mix/lib/mix/local.ex#L144)?

I am confused how this wouldn't apply to downloads from `cdn.jsdelivr.net/hex`. I had a quick look at the code, which obviously you contributed to, and it seems to me that mix also verifies the signature with the mirror by using signature csv files for [hex](http://cdn.jsdelivr.net/hex/installs/hex-1.x.csv) and for [rebar](https://cdn.jsdelivr.net/hex/installs/rebar3-1.x.csv).

> It is also known to sometimes lag behind by a few days on certain paths.

In the current implementation, for Linux OTP and Elixir builds, the action first tries to fetch the list of builds and it does so in the order of the mirrors, so starting with the official mirror. And then it proceeds with downloading the binary, proceeding likewise in the order of mirrors. The only case where it would pick an older version (compared to the official mirror) would be if the official mirror didn't respond properly.

For consumers of the action, do we want by default to fail a build because `repo.hex.pm` is not available or fails (which it did a lot recently) without trying next with a mirror that might save the build?

The whole point of this PR is to make builds that depend on this action more robust. It seems to me that defaulting to a maybe lagging mirror is a good solution. If consumers want a particular version, they can mention it in the spec and if this version is not available in the mirror, the build will obviously fail.

Reply:
> Are you referring to the mechanism in [Mix.Local](https://github.com/elixir-lang/elixir/blob/d9c7e80302c220a2b603ac16931dee6056a6c863/lib/mix/lib/mix/local.ex#L144)?

`mix local.hex` and `mix local.rebar` will be verified but the Elixir and Erlang installations are not installed through Mix and will not be verified.

> For consumers of the action, do we want by default to fail a build because repo.hex.pm is not available or fails (which it did a lot recently) without trying next with a mirror that might save the build?

I think we should retry our fetching from repo.hex.pm X number of times. We don't know if trying another mirror will be more successful than retrying the original mirror. I don't think we should default to falling back to a mirror we have no control over, can provide no guarantees for, and we cannot verify the integrity of the files we dowload.

Adding support for additional mirrors is great but if we pick an unofficial mirror to use as default I would like to see integrity checks for the files we downloads and stats that show that retrying to it will have a higher success rate.

Reply:
I agree with @ericmj.

The **default**, if the option is not used, should be whatever behaviour we had before (except for the obvious cache).

The fact is nobody else (up until now, at least) had similar issues (that were reported), so the option itself should be opt-in.

On the other hand, if you do decide to opt-in maybe 3 tries, with exponential fallback would be Ok in a round-robin fashion. What do you think, @ericmj?

e.g. 

```
hexpm-mirrors:
  - A
  - B
  - C
```

would either do A, A, A, B, B, B, C, C, C, or A, B, C, A, B, C, ... with exponential backoff every time a previous mirror is retried.

Also, @pguyot, it would be nice to add a "how to use it" to the README. I don't expect consumers of the action to have to go to the code for that. There's an example section at the bottom that might be appropriate for an example.

Reply:
I think `A, A, A, B, B, B, C, C, C` should be the behavior.

Reply:
We dropping retries, and changing the implemented mechanism, so this conversation is no longer relevant.